Камень судьбы

nerodemiurgo opened the pull request:
Перечислены константы действий и статусов,
текущий статус и идентификаторы пользователей сохранены в свойствах,
описаны методы передачи действий и статусов,
описан метод вычисления действий для текущего статуса и роли,
описан метод вычисления следующего статуса по пришедшему действию,
убраны break перед return

---
:mortar_board: [Камень судьбы](https://up.htmlacademy.ru/yii/2/user/336921/tasks/3)

nerodemiurgo committed:
Перечислены константы действий и статусов,
текущий статус и идентификаторы пользователей сохранены в свойствах,
описаны методы передачи действий и статусов,
описан метод вычисления действий для текущего статуса и роли,
описан метод вычисления следующего статуса по пришедшему действию

nerodemiurgo committed:
Перечислены константы действий и статусов,
текущий статус и идентификаторы пользователей сохранены в свойствах,
описаны методы передачи действий и статусов,
описан метод вычисления действий для текущего статуса и роли,
описан метод вычисления следующего статуса по пришедшему действию

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+        $this->role = $role;
+    }
+
+// Определять список из всех доступных действий
```
Комментарии к методам нужно оформлять как phpDoc
```suggestion
    /**
     * Определять список из всех доступных действий
     */
```

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+    }
+// Определять список из всех доступных статусов
+    public function getAllStatuses() {
+        $statuses = array_keys(self::STATUSES);
```
Тут вернётся 0,1,2....

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+    public $role = "";                            //Роль пользователя, совершившего действий
+
+    const STATUSES = [
+        'new',                    //Статус новое
```
Нужно каждое значение вынести в отдельную константу: `const STATUS_NEW = 'new'`

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+    ];
+
+    const ACTIONS = [
+       'to_cancel'       => ['cancel'],             //Отменить заказ
```
Каждое действие тоже вынести в константу

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+
+class Task
+{
+    public $current_status = "";                  //Текущий статус, обязательное
+    public $action = "";                          //Пришедшее действие
+    public $customer_id = 0;                      //ID заказчика, обязательное
+    public $executor_id = 0;                      //ID исполнителя, обязательноe
+    public $role = "";                            //Роль пользователя, совершившего действий
```
Это всё стоит сделать private, значения по умолчанию не нужны, т.к. задаются в конструкторе

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+class Task
+{
+    public $current_status = "";                  //Текущий статус, обязательное
+    public $action = "";                          //Пришедшее действие
```
Это не нужно

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+    }
+// Определять список доступных действий в текущем статусе
+    private function getActions($current_status, $role) {
+        if ($role = "customer") {
```
Роли тоже вынеси в константы

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+    public function getNextStatus($action) {
+        switch ($action) {
+            case self::ACTIONS['to_cancel']:
+                return self::STATUSES['canceled'];
```
Таких ключей там нет

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,78 @@
+                return self::STATUSES['new'];
+        }
+    }
+}
```
Нужно ещё написать файл test.php, в котором проверить результат вызова getActions и getNextStatus

nerodemiurgo committed:
- изменен вид комментариев
- константы выведены в отдельное значение
- свойства сделаны private
- свойства текущего статуса и действия убраны
- роли вынесены в константы
- добавлен test.php

nerodemiurgo committed:
- изменен вид комментариев
- константы выведены в отдельное значение
- свойства сделаны private
- свойство текущего действия убрано
- роли вынесены в константы
- добавлен test.php

b1rdex commented on test.php:
```diff
@@ -0,0 +1,23 @@
+$getStatuses = $task->getAllStatuses();
+print("<h2>Все значения статусов</h2>");
+foreach($getStatuses as $s => $value) {
+    print($s." - ".$value."<br>");
```
В тестах должны быть какие-то проверки результатов, как вариант через assert()
Чтобы если появится ошибка в логике, тест завершался ошибкой

b1rdex replied:
Т.е. это не ручная проверка результатов print глазами

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,103 @@
+    const ROLE_CUSTOMER = "customer";
+    const ROLE_EXECUTOR = "executor";
+
+    public function __construct($current_status, $customer_id, $executor_id, $role)
```
По доп. критериям всем аргументам всех методов нужно проставлять типы

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Определять список из всех доступных действий
+     **/
+    public function getAllActions()
```
И возвращаемые типы всем методам тоже

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,103 @@
+    /**
+     * Определять список доступных действий в текущем статусе
+     **/
+    private function getActions($current_status, $role)
+    {
+        if ($this->role == self::ROLE_CUSTOMER) {
```
Стоит удалить $this->role и тут ты правильно добавила $role аргументом. Ведь по логике вещей, у задачи не может быть внутреннего состояния Роль. Но в момент выполнения метода getActions мы знаем для кого его вызываем и поэтому передаём $role.

b1rdex commented on Task.php:
```diff
@@ -0,0 +1,103 @@
+    private function getActions($current_status, $role)
+    {
+        if ($this->role == self::ROLE_CUSTOMER) {
+            switch ($current_status) {
```
А с $current_status ситуация наоборот – это внутреннее состояние и мы уже храним его в $this->current_status. Поэтому аргумент $current_status нужно удалить и исправить условие

b1rdex commented on test.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+
+require_once'Task.php';
```
По критериям код нужно форматировать в соответствии с PSR-2

b1rdex commented on test.php:
```diff
@@ -0,0 +1,23 @@
+}
+
+print("<h2>Тест определения следующего статуса после конкретного действия</h2>");
+$action = "take_to_work";
```
Используй константы, для этого мы их и делали :)
```suggestion
$action = Task::ACTION_TO_TAKE_TO_WORK;
```

nerodemiurgo committed:
Изменено имя наставника.

nerodemiurgo committed:
Тестовый сценарий переписан под assert.
Проставлены типы аргументам и возвращаемым методам.

nerodemiurgo committed:
Тестовый сценарий переписан под assert.
Проставлены типы аргументам и возвращаемым методам.
Убран лишний аргумент метода getActions

sterlikoff reviewed (changes requested):
Ольга, в целом все хорошо, но есть пара моментов, на которые стоит обратить внимание:

1. Разработку следовало начать с написания интерфейса, от которого уже наследовать класс, в котором методы интерфейса реализуются. Наверное стоит повторить "Знакомство с ООП".

2. В методе getActions есть кейс, в котором функция может вообще ничего не вернуть (если role и не customer и не executor), а заявлено всегда string. Да, это как будто бы невозможно (с нашей точки зрения, а с точки зрения php - нет). А наша точка зрения может поменяться в ходе изменения ТЗ, например.

3. В тестах используешь нестрогие сравнения (==), старайся использовать строгие (===)

4. Старайся избегать автоприведения типов в assert так, чтобы первым аргументом всегда был bool:

assert(!empty($task->getAllStatuses()), 'Перечень статусов не возвращается');

nerodemiurgo committed:
Исправлены названия методов, добавлен return в метод getActions, поправлены ассерты.

sterlikoff reviewed (changes requested):
Отлично, почти все :) 

Все еще хочу увидеть в классе наследование от интерфейса и еще обратил внимание, что `getActions` не участвует в тестах. 

Поправь пожалуйста это для зачета :)

nerodemiurgo committed:
Метод getActions сделан публичным и добавлен в тесты.

sterlikoff reviewed (approved):
Ольга, теперь все отлично :) Зачет, идем дальше!
